packages/fxa-auth-server/lib/redis.js

```diff
@@ -4,37 +4,106 @@
 
 'use strict';
 
-const error = require('./error');
+const Redis = require('ioredis');
+const { readdirSync, readFileSync } = require('fs');
+const { basename, extname, resolve } = require('path');
 
-module.exports = (config, log) => {
-  const redis = require('../../fxa-shared/redis')(config, log);
-  if (!redis) {
-    return;
+const scriptNames = readdirSync(resolve(__dirname, 'luaScripts'), {
+  withFileTypes: true,
+})
+  .filter(de => de.isFile() && extname(de.name) === '.lua')
+  .map(de => basename(de.name, '.lua'));
+
+function readScript(name) {
+  return readFileSync(resolve(__dirname, 'luaScripts', `${name}.lua`), {
+    encoding: 'utf8',
+  });
+}
+
+class FxaRedis {
+  constructor(config, log) {
+    config.keyPrefix = config.prefix;
+    this.log = log;
+    this.redis = new Redis(config);
+    scriptNames.forEach(name => this.defineCommand(name));
+  }
+
+  defineCommand(name, numberOfKeys = 1) {
+    this.redis.defineCommand(name, {
+      lua: readScript(name),
+      numberOfKeys,
+    });
+  }
+
+  touchSessionToken(uid, token) {
+    return this.redis.touchSessionToken(uid, JSON.stringify(token));
   }
 
-  return Object.entries(redis).reduce((object, [key, value]) => {
-    if (typeof value === 'function') {
-      object[key] = async (...args) => {
-        try {
-          return await value(...args);
-        } catch (err) {
-          if (err.message === 'redis.watch.conflict') {
```

Comment:
Wahey, no more conflicts! I assume this is because of [redis' guarantee of the atomicity of scripts](https://redis.io/commands/eval#atomicity-of-scripts) and this snippet: `Redis guarantees that a script is executed in an atomic way: no other script or Redis command will be executed while a script is being executed.`

Reply:
precisely